bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -0,0 +1,230 @@
+    }
+  }
+
+
+  private def addProgram(prog: ISPProgram, obsList: List[ISPObservation]): Unit = {
+    import LchQueryParam.{ISPProgramExtractors, ISPObservationExtractors, ToYesNo}
+
+    def makeTargetNode(target: SPTarget, env: TargetEnvironment): Option[Serializable] = {
+      def targetType: Option[String] = {
+        if (env.isBasePosition(target)) Some("Base")
+        else if (env.isUserPosition(target)) Some("User")
+        else if (env.isGuidePosition(target)) {
+          for {
+            gg  <- env.getGroups.asScala
+            gpt <- gg.getAll.asScala
+            if gpt.containsTarget(target)
+          } yield gpt.getGuider.getKey
+        }.headOption
+        else None
+      }
+
+      if (target.isNonSidereal) {
+        Some(new NonSidereal() {
+          setName(target.getName)
+          setType(targetType.orNull)
+        })
+      } else if (target.isSidereal) {
+        target.getSkycalcCoordinates(None.asGeminiOpt).asScalaOpt.map { coords =>
+          new Sidereal() {
+            setName(target.getName)
+            setType(targetType.orNull)
+            setHmsDms(new HmsDms() {
+              setRa(HHMMSS.valStr(coords.getRa.getMagnitude))
+              setDec(DDMMSS.valStr(coords.getDec.getMagnitude))
+            })
+          }
+        }
+      } else None
+    }
+
+
+    def makeTimingWindow(spTW: SPSiteQuality.TimingWindow): TimingWindow = {
+      def timingString(ms: Long) =
+        f"${ms / LchQueryFunctor.MsPerHour}%d:${(ms % LchQueryFunctor.MsPerHour) / LchQueryFunctor.MsPerMinute }%02d"
+
+      def formatDuration: Option[String] =
+        Option(spTW.getDuration).
+          filterNot(_ == SPSiteQuality.TimingWindow.WINDOW_REMAINS_OPEN_FOREVER).
+          map(timingString)
+
+      def formatWindow: String =
+        LchQueryFunctor.dateFormat.format(new Date(spTW.getStart))
+
+      def formatPeriod: Option[String] =
+        Option(spTW).
+          filterNot(_.getDuration == SPSiteQuality.TimingWindow.WINDOW_REMAINS_OPEN_FOREVER).
+          filterNot(_.getRepeat == SPSiteQuality.TimingWindow.REPEAT_NEVER).
+          map(t => timingString(t.getPeriod))
+
+      def formatTimes: Option[String] =
+        Option(spTW).
+          filterNot(_.getDuration == SPSiteQuality.TimingWindow.WINDOW_REMAINS_OPEN_FOREVER).
+          filterNot(_.getRepeat == SPSiteQuality.TimingWindow.REPEAT_FOREVER).
+          filterNot(_.getRepeat == SPSiteQuality.TimingWindow.REPEAT_NEVER).
+          map(tw => String.valueOf(tw.getRepeat))
+
+      new TimingWindow() {
+        setDuration(formatDuration.orNull)
+        setTime(formatWindow)
+        if (spTW.getRepeat != SPSiteQuality.TimingWindow.REPEAT_NEVER) {
+          setRepeats(new TimingWindowRepeats() {
+            setPeriod(formatPeriod.orNull)
+            setTimes(formatTimes.orNull)
+          })
+        }
+      }
+    }
+
+    queryResult.getProgramsNode.getPrograms.add {
+      val spProg = prog.toSPProg.get
+      new Program() {
```

Comment:
Yes... all the classes used internally in `LchQueryFunctor`, I believe, are from an LCH library.